### PR TITLE
tests: mem_protect: add a test case of adding memory partition

### PR DESCRIPTION
**Pull request by maksimmasalski**
Add a test case to validate when adding a new partition into a memory
domain with over its maximum specified limit number, an assertion
failure happens.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

### CODE REVIEW

**Review by enjiamai — COMMENTED** [review]
Excuse me, one question for this, how do we judge the test case success or not, when the assertion is happened,  don't we have to something for this? 
or the assertion(message) will be captured by test harness(in testcase.yaml file), 
please help to figure out my concern, thanks~

**Comment by enjiamai** [comment]
About above question, is it's verified mechanism the same as you reply to Steven in #25535,  test case: void test_mem_domain_api_kernel_thread_only(void)?

**Comment by maksimmasalski** [comment]
> About above question, is it's verified mechanism the same as you reply to Steven in #25535, test case: void test_mem_domain_api_kernel_thread_only(void)?

Yes, it is.

**Review by enjiamai — APPROVED** [review]

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c [hunk 1]
```diff
@@ -481,6 +487,29 @@ static void mem_domain_test_6_2(void *p1, void *p2, void *p3)
 	zassert_unreachable(ERROR_STR);
 }
 
+/**
+ * @brief Test system assert when adding memory partitions more then possible
+ *
+ * @details
+ * - Add memory partitions one by one and more then architecture allows to add.
```
Comment (LeiW000) [inline tests/kernel/mem_protect/mem_protect/src/mem_domain.c:494]: I guess that "then" should be "than".

**Review by LeiW000 — CHANGES_REQUESTED** [review]

**Review by LeiW000 — APPROVED** [review]

**Comment by maksimmasalski** [comment]
@carlescufi Hello, Carles. Could you merge that PR?

**Comment by MaureenHelm** [comment]
@andrewboie will you please review?

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c [hunk 2]
```diff
@@ -379,3 +379,28 @@ void test_mem_part_overlap(void)
...
+ */
+void test_mem_part_assert_add_overmax(void)
+{
+	int max_parts;
+
+	set_fault_valid(true);
+
+	max_parts = arch_mem_domain_max_partitions_get();
+	for (int i = 0; i <= max_parts; i++) {
+		k_mem_domain_add_partition(&test_domain, &part_max);
```
Comment (andrewboie) [inline tests/kernel/mem_protect/mem_protect/src/mem_domain.c:404]: this isn't testing what you think it tests.
memory domain partitions are documented as not being allowed to overlap, so it should be asserting as soon as you try to add the same partition a second time.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c (hunk 2 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/mem_domain.c:404]: In another merged PR I already created a test to check system assert when partitions overlap #30004 
Are you sure that we are not talking about that PR #30004 ?

In the current PR the goal is to add partitions more than maximum allowed and to capture a correct error.

Comment (maksimmasalski) [reply tests/kernel/mem_protect/mem_protect/src/mem_domain.c:404]: @andrewboie Could you take a look at?

**Comment by LeiW000** [comment]
@andrewboie , could you help to take look at the PR again.

**Comment by LeiW000** [comment]
@enjiamai , could you help to address the conflict?

**Comment by enjiamai** [comment]
> @enjiamai , could you help to address the conflict?

Yes, I just did that~

**Comment by enjiamai** [comment]
@LeiW000  It seems like I did it wrong......sorry, let me see how to address that.

**Comment by enjiamai** [comment]
Sorry, I think I screw it up, the best solution might be:  @maksimmasalski  could you please rebase it, thank you!

**Comment by LeiW000** [comment]
@enjiamai, do we still need Maksim's help to rebase the code?

**Comment by enjiamai** [comment]
> @enjiamai, do we still need Maksim's help to rebase the code?

@LeiW000 Rebasing is needed, I will contact @maksimmasalski  by IM today.

**Comment by nashif** [comment]
I am not even sure what 'when overmax add partitions' means here...

**Comment by enjiamai** [comment]
> I am not even sure what 'when overmax add partitions' means here...

Hi, thanks for correcting this, I make some update on this.

**Comment by enjiamai** [comment]
@dcpleung , could you please also help to review this one? thanks!

**Comment by ioannisg** [comment]
@maksimmasalski @dcpleung it is not clear to me whether the comment by @andrewboie regarding what is actually tested, is addressed. It seems that the kernel will assert anyway, if we try to add the same partition multiple ties regardless of whether we exceed the number of partitions supported by the arch

**Comment by ioannisg** [comment]
@nashif i suggest we do not merge this until we are certain that the -1 from @andrewboie is addressed.

**Review by ioannisg — COMMENTED** [review]
Not sure if the feedback by @andrewboie is addressed.

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c [hunk 3]
```diff
@@ -48,9 +48,9 @@ void test_mem_domain_setup(void)
 	struct k_mem_partition *parts[] = { &ro_part, &ztest_mem_partition };
 
 	num_rw_parts = max_parts - PARTS_USED;
-	zassert_true(NUM_RW_PARTS >= num_rw_parts,
-		     "CONFIG_MAX_DOMAIN_PARTITIONS incorrectly tuned, %d should be at least %d",
-		     CONFIG_MAX_DOMAIN_PARTITIONS, max_parts);
+	zassert_true(num_rw_parts <= NUM_RW_PARTS,
+			"CONFIG_MAX_DOMAIN_PARTITIONS incorrectly tuned, %d should be at least %d",
+			CONFIG_MAX_DOMAIN_PARTITIONS, max_parts);
```
Comment (dcpleung) [inline tests/kernel/mem_protect/mem_protect/src/mem_domain.c:53]: This is functionally the same. Why is this being changed?

**Review by dcpleung — CHANGES_REQUESTED** [review]
Comment from @andrewboie still stands. The assertion is about the same partition being added multiple times (=> overlapping).

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c (hunk 3 above)
Comment (enjiamai) [reply tests/kernel/mem_protect/mem_protect/src/mem_domain.c:53]: Hi @dcpleung,  @maksimmasalski  was trying to fix the warning message of checkpatch.pl:
```
tests/kernel/mem_protect/mem_protect/src/mem_domain.c:51: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#51: FILE: tests/kernel/mem_protect/mem_protect/src/mem_domain.c:51:
+       zassert_true(NUM_RW_PARTS >= num_rw_parts,

tests/kernel/mem_protect/mem_protect/src/mem_domain.c total: 0 errors, 1 warnings, 483 lines checked
```
So I think this change acceptable, right? thanks!

**Comment by enjiamai** [comment]
> Not sure if the feedback by @andrewboie is addressed.

Hi @ioannisg , I think the question from @andrewboie has been addressed, @maksimmasalski  has updated it. And this test case aims to add a memory partition over the max limit,  and it has a different purpose from the previous one (add a memory partition that overlaps the existing one.). And thanks for your review!

**Comment by LeiW000** [comment]
@dcpleung, Andrew's comment has been addressed. Could you take look at the PR again? We would like to get it merged before the feature freeze(05/07).

**Comment by enjiamai** [comment]
> Comment from @andrewboie still stands. The assertion is about the same partition being added multiple times (=> overlapping).

Sorry for missing the message, I check it in detail, yes, the exception is due to overlapping, not exceed the max partition. I will update this soon. thanks!

File: tests/kernel/mem_protect/mem_protect/src/mem_domain.c (hunk 2 above)
Comment (enjiamai) [reply tests/kernel/mem_protect/mem_protect/src/mem_domain.c]: sorry, @ioannisg, @dcpleung, I had some misunderstanding and redesign this case,  the comment from Andrewboie has been addressed in this update. Thanks so much for correcting this.

**Comment by enjiamai** [comment]
Thanks, @ioannisg @dcpleung , sorry for bugging you again. My understanding is not correct previously. So I redesign this case and I think this time the comment has been addressed, could you please help to review it again?

**Review by dcpleung — APPROVED** [review]

**Pull request merged**